IQSS/dataverse-frontend
# Multiple dynamic fields - Create Dataset Form

**What this PR does / why we need it**:
Adds the ability to add multiple values to a field.

**Which issue(s) this PR closes**:

Closes #376

**Special notes for your reviewer**:
- New **_Select Multiple_** component integrated in the form to replace the multiple checkbox.
- _**FormGroupWithMultipleFields**_  is no longer responsible for adding or removing elements, that is taken care of by the consumer logic (SPA).
- I need to remove Form Group `required` and `withinMultipleFields` props being cloned to all childrens inside because did not work correctly when we had nested elements to clone within the Forms Group (e.g. adding props to all elements unnecessarily).
- I removed `withinMultipleFields` props from form elements and with it the `<FormElementLayout/>`, so now the design system is not responsible for rendering the form group as a column or as a row.
In some cases within the Create Dataset form when working on this issue I found some complications when working with so many dynamic fields.

I think it's pretty useful to just pass a prop to render the form group one way or another, but I think the use of `react.cloneElement` was being counter-intuitive in this case.
Perhaps we could see in a separate issue a way to accomplish the same thing but without the use of `cloneElement`.


**Suggestions on how to test this**:
### Step 1: Run the Development Environment
1. Execute `npm i`.
2. Navigate with `cd packages/design-system && npm run build`.
3. Return with `cd ../../`.
4. Ensure you have a `.env` file similar to `.env.example`, with the variable `VITE_DATAVERSE_BACKEND_URL=http://localhost:8000`.
5. Navigate with `cd dev-env`.
6. Start the environment using `./run-env.sh unstable`.
7. To verify the environment, visit <http://localhost:8000> and check your local Dataverse installation.

### Step 2: Test the Create Dataset Form
- Log in, click the create dataset button and the form fields should be displayed successfully.
- Add or remove a field by clicking the buttons to the right of it.
- Complete all required fields and save the form, everything should keep working as usual.
- Try closing the accordion and clicking the Save button without filling any field, the accordion with errors should open and scroll and focus to the first field with errors.

**Is there a release notes update needed for this change?**:

**Does this PR introduce a user interface change? If mockups are available, please link/include them here:
Yes.
![DynamicMultipleFields](https://github.com/IQSS/dataverse-frontend/assets/64560524/00b213ec-bc76-44f3-9d0d-9282f16c4531)

**Additional documentation**:
No.
